src/setup.jl

```diff
@@ -21,7 +21,8 @@ function _initialize_subproblems(sp_map::Dict{Int,Set{ScenarioID}},
                                  constructor::Function,
                                  constructor_args::Tuple,
                                  r::AbstractPenaltyParameter,
-                                 warm_start::Bool;
+                                 warm_start::Bool,
+                                 subproblem_callbacks::Vector{SubproblemCallback};
```

Comment:
Yes, I use tabs. I have to admit I don't know how you format PH.jl - I use VSCode and it doesn't seem to work well with your formatting, perhaps there's a plugin I should be using?

Reply:
I mostly just go by what "looks good" which is totally subjective and unhelpful.  But some rough guidelines:

1. Use whitespace (blank lines, separating long lines, etc.) liberally.  This makes the code easier on the eyes.
2. A single line shouldn't be more than roughly 80 characters (with the notable exception of string literals).
3. If a function signature or call doesn't fit on a single line break it up with the first argument following the function, and every other argument on its own line. I've started doing this in the last year. It is not consistently done throughout PH because of that.  Feel free to update areas with this guideline.
4. Within a file, group functions in "logical" groups. Try to alphabetize within a group.  The most obvious groupings are internal functions (i.e. not meant to be exposed to the user--normally preceeded by an "_" in the name) and external functions.

I hope that helps a little bit. I should probably post some sort of guidelines but before your contributions, I was the only one working in this code and it wasn't necessary.

There does seem to be a julia extension for VSCode: https://github.com/julia-vscode/julia-vscode

I just googled "julia vscode" and. that was what first came up.  There may be other options you like better.

Reply:
I mostly follow 1, 2 and 3, but my formatting is usually slightly different, e.g.
```julia
    param = Dict(
        "optimizer" => get_optimizer(sim),
        "relativePathTimeSeriesCSV" => ts_file,
        "optimizationHorizon" => Dict(
            "start" => [1,1,1],
            "end" => [
                1,
                length(sim["years to consider"])*length(sim["periods to consider"]),
                sim["time steps per period"]
            ]
        ),
    )
```
or
```julia

function build_GEP_scenario(
        scenario_id::PH.ScenarioID, 
        param::Dict,
        configFile,
        sim::Simulation,
        geppr_ts_file::String
    )
```
Atom (my previous editor) and VSCode allow for this quite easily, whereas your formatting is a little less convenient since I have to tab and then space to the opening bracket `(`, and similarly for the closing bracket.

I'll try to follow your guideline also for 4 next time I work on this tomorrow. On a sidenote, I think that `test_subproblem_callbacks` is probably unnecessary, since it's just two tests.

Reply:
I use emacs which does both automatically for me.  Your first example is actually exactly why I use this formatting.  I find it significantly easier to see what's going on with this formatting
```julia
param = Dict("optimizer" => get_optimizer(sim),
             "relativePathTimeSeriesCSV" => ts_file,
             "optimizationHorizon" => Dict("start" => [1,1,1],
                                           "end" => [1,
                                                     length(sim["years to consider"])*length(sim["periods to consider"]),
                                                     sim["time steps per period"]
                                                     ]
                                           ),
             )
```
even with the line length.

Admittedly, its just because I've been looking at code like that for a long time and I'm just used to it.

Reply:
Found what I was looking for :-) https://marketplace.visualstudio.com/items?itemName=LAK132.indent-to-bracket